### PR TITLE
Add Twitter Support

### LINKED ISSUE #51: Add Twitter/Birdsong
We want to migrate Twitter to here from Zenodotus. That's mostly so they're all in one place.

### PR DESCRIPTION
**Pull request by cguess**
This adds twitter support via the Birdsong gem. This is designed to move
twitter out of Zenodotus and into the regular pipeline, giving us better
separations of concerns.

To test:
- Update your version of selenium server to 4.4.0 from https://www.selenium.dev/downloads/
- `bundle install`
- Start the selenium server `java -jar ./selenium-server-4.4.0.jar standalone --session-timeout 1000`
- Start up sidekiq/redis etc
- `rails t`

Closes #51

### CODE REVIEW

File: lib/libraries/aws_object_upload_file_wrapper.rb [hunk 1]
```diff
@@ -19,6 +19,9 @@ def initialize(file_path)
   # @return [Boolean] True when the file is uploaded; otherwise false.
   def upload_file
     @object.upload_file(@file_path)
+
+    # Delete the image after it's uploaded
+    File.delete(@file_path) if File.exist? @file_path
```
Comment (oneroyalace) [inline lib/libraries/aws_object_upload_file_wrapper.rb:24]: 👍🏿

**Review by oneroyalace — COMMENTED** [review]
Running tests now

File: app/media_sources/twitter_media_source.rb [hunk 2]
```diff
@@ -0,0 +1,138 @@
...
+  # @!scope class
+  # @params url [String] the url of the page to be collected for archiving
+  # @params save_screenshot [Boolean] whether to save the screenshot image (mostly for testing).
+  #   Default: false
+  # @returns [String or nil] the path of the screenshot if the screenshot was saved
+  def self.extract(scrape, save_screenshot = false)
+    object = self.new(scrape.url)
+    object.retrieve_tweet
+  rescue StandardError => error
+    error_message = "*Zorki Error 📸:*\n`#{error.class.name}`\n> #{error.message}\n*URL Submitted:* #{scrape.url}"
```
[the remark this answers is not in the document]
Comment (cguess) [reply app/media_sources/twitter_media_source.rb:30]: Bleh. This was broken in Zeno too. WHOOPS

File: app/media_sources/twitter_media_source.rb [hunk 3]
```diff
@@ -0,0 +1,138 @@
...
+  # @return [Birdsong::Tweet]
+  def retrieve_tweet
+    id = TwitterMediaSource.extract_tweet_id_from_url(@url)
+    tweet = Birdsong::Tweet.lookup(id).first
+
+    self.class.create_aws_key_functions_for_posts([tweet])
+
+    return tweet unless s3_transfer_enabled?
+
+    @@logger.debug "Beginning uploading of files to S3 bucket #{Figaro.env.AWS_S3_BUCKET_NAME}"
```
[the remark this answers is not in the document]
Comment (cguess) [reply app/media_sources/twitter_media_source.rb:61]: Yes, this is done in the gem however. So good point, and I'm going to do that first then add it to this PR.

File: app/blueprints/twitter_post_blueprint.rb [hunk 4]
```diff
@@ -0,0 +1,58 @@
...
+          :text,
+          :language,
+          :created_at
+
+  association :author, blueprint: TwitterUserBlueprint
+
+  field :object_type do |tweet|
+    "twitter"
+  end
+
```
[the remark this answers is not in the document]
Comment (cguess) [reply app/blueprints/twitter_post_blueprint.rb:14]: See the previous comment.

File: app/media_sources/media_source.rb [hunk 5]
```diff
@@ -53,11 +53,13 @@ def self.take_screenshot(url: @url, indicator_element_id: "")
     session.visit(url)
     begin
       session.find_by_id(indicator_element_id) # Block until page content loadsrescue
-    rescue Capybara::ElementNotFound
-    end
+    rescue Capybara::ElementNotFound; end
+
     media_source_name = self.to_s.delete_suffix("MediaSource").downcase
-    screenshot_path = session.save_screenshot("/tmp/#{media_source_name}_screenshot_#{SecureRandom.uuid}.png")
+    save_path = File.join(Rails.root, "tmp", "#{media_source_name}_screenshot_#{SecureRandom.uuid}.png")
```
Comment (oneroyalace) [inline app/media_sources/media_source.rb:59]: 👍🏿 much nicer

File: test/media_sources/twitter_media_source_test.rb [hunk 6]
```diff
@@ -43,6 +43,15 @@ def test_initializing_returns_blank
     tweets.each { |tweet| assert_not_nil(tweet.aws_image_keys) }
   end
 
+  test "tweet has screenshot aws key" do
+    skip unless ENV["AWS_REGION"].present?
+
+    tweets = TwitterMediaSource.extract(Scrape.create({ url: "https://twitter.com/Space4Europe/status/1552221138037755904" }))
+    assert_not_nil(tweets.first.aws_screenshot_key)
+
+    tweets.each { |tweet| assert_not_nil(tweet.aws_image_keys) }
```
Comment (oneroyalace) [inline test/media_sources/twitter_media_source_test.rb:52]: I believe this should be `tweet.aws_screenshot_key`, right?

Comment (cguess) [reply test/media_sources/twitter_media_source_test.rb:52]: No, `tweets` is an array so I'm taking the first element.

Comment (oneroyalace) [reply test/media_sources/twitter_media_source_test.rb:52]: Er. nvm.

**Comment by oneroyalace** [comment]
I saw this error while testing: 
```
Error:
ScraperControllerTest#test_scraping_a_twitter_image_with_force_works:
NoMethodError: undefined method `has_key?' for ["id", "1562462774969581570"]:Array

      assert JSON.parse(JSON.parse(@response.body)["scrape_result"]).first.has_key?("id")
                                                                          ^^^^^^^^^
    test/controllers/scraper_controller_test.rb:162:in `block (2 levels) in <class:ScraperControllerTest>'
    test/controllers/scraper_controller_test.rb:159:in `block in <class:ScraperControllerTest>'
```

**Comment by cguess** [comment]
Hrm... I cannot recreate these errors. Can you make sure you can access the url that it's trying to scrape?

**Comment by reefdog** [comment]
Here's what I'm getting:

<details><summary>/Users/justin/.rbenv/versions/3.1.2/lib/ruby/gems/3.1.0/bundler/gems/zorki-a53f7cf00e92/lib/zorki/scrapers/scraper.rb:71:in `get_content_of_subpage_from_url': Zorki could not find content requested (Zorki::ContentUnavailableError)</summary>

```
➜  hypatia git:(20220727-add-twitter) rails t                                                               
Calling `DidYouMean::SPELL_CHECKERS.merge!(error_name => spell_checker)' has been deprecated. Please call `DidYouMean.correct_error(error_name, spell_checker)' instead.
I, [2022-08-26T11:13:58.979483 #44745]  INFO -- : /Users/justin/.rbenv/versions/3.1.2/lib/ruby/gems/3.1.0/bundler/gems/zorki-a53f7cf00e92/lib/zorki/monkeypatch.rb is monkeypatching Selenium::WebDriver::DevTools.send_cmd
I, [2022-08-26T11:14:31.573749 #44745]  INFO -- : Forki download started: https://scontent-dfw5-1.xx.fbcdn.net/v/t39.30808-6/272678720_10161587852463065_4418883591852921526_n.jpg?_nc_cat=110&ccb=1-7&_nc_sid=730e14&_nc_ohc=_-U-9IrVnB8AX-vCRSR&_nc_ht=scontent-dfw5-1.xx&oh=00_AT_zHmrtNdHkxQwpD9zy5C4VMWjRq_3l6QSCqC3cFUYZLw&oe=630ED25D
I, [2022-08-26T11:14:31.768877 #44745]  INFO -- : Forki download finished
I, [2022-08-26T11:14:31.768944 #44745]  INFO -- : Save Location: tmp/forki/facebook_media_9afd914e-c062-49bf-a218-c5928532c29a.jpg
I, [2022-08-26T11:14:31.769013 #44745]  INFO -- : Size: 0.3321 MB
I, [2022-08-26T11:14:31.769065 #44745]  INFO -- : Time to Download: 0.195 seconds
I, [2022-08-26T11:14:57.459611 #44745]  INFO -- : Forki download started: https://scontent-dfw5-2.xx.fbcdn.net/v/t1.6435-1/131583430_10160550552578065_3060525849019065048_n.jpg?stp=dst-jpg_p480x480&_nc_cat=107&ccb=1-7&_nc_sid=7206a8&_nc_ohc=5KPZOw7THVAAX_gGp0z&_nc_ht=scontent-dfw5-2.xx&oh=00_AT9hPTShwzuRAa8T-NqxdVLmKlg0nY4sn337MfewvtkYXw&oe=632F0790
I, [2022-08-26T11:14:57.590901 #44745]  INFO -- : Forki download finished
I, [2022-08-26T11:14:57.590952 #44745]  INFO -- : Save Location: tmp/forki/facebook_media_a5658c51-02ca-42e6-972f-861daee57dd0.jpg
I, [2022-08-26T11:14:57.590976 #44745]  INFO -- : Size: 0.0392 MB
I, [2022-08-26T11:14:57.590996 #44745]  INFO -- : Time to Download: 0.131 seconds
I, [2022-08-26T11:16:33.379845 #44745]  INFO -- : Forki download started: https://scontent-dfw5-1.xx.fbcdn.net/v/t15.5256-10/256109747_264439672183864_4303307282486204364_n.jpg?_nc_cat=109&ccb=1-7&_nc_sid=df419e&_nc_ohc=1DabxSuZbjkAX9OOclF&_nc_ht=scontent-dfw5-1.xx&oh=00_AT9HVtndWS6iq-m6Aj80FskhrLVLqYYlzrLDgFCaVSA-sQ&oe=630E5B2D
I, [2022-08-26T11:16:33.550320 #44745]  INFO -- : Forki download finished
I, [2022-08-26T11:16:33.550382 #44745]  INFO -- : Save Location: tmp/forki/facebook_media_ca3f6ea8-3394-47d8-9bc7-b3a1cf0d0812.jpg
I, [2022-08-26T11:16:33.550403 #44745]  INFO -- : Size: 0.1807 MB
I, [2022-08-26T11:16:33.550425 #44745]  INFO -- : Time to Download: 0.171 seconds
I, [2022-08-26T11:16:33.550434 #44745]  INFO -- : Forki download started: https://scontent-dfw5-2.xx.fbcdn.net/v/t42.1790-4/248023383_676389286657332_158189225395188557_n.mp4?_nc_cat=106&ccb=1-7&_nc_sid=985c63&efg=eyJ2ZW5jb2RlX3RhZyI6InN2ZV9zZCJ9&_nc_ohc=k_UyPyZki8MAX-xQr7-&_nc_ht=scontent-dfw5-2.xx&oh=00_AT-GDRuGockVxwXaWlvXjsXdlOymSz2qlt0Uu2PTc_oDCA&oe=63090FE8
I, [2022-08-26T11:16:33.895157 #44745]  INFO -- : Forki download finished
I, [2022-08-26T11:16:33.895192 #44745]  INFO -- : Save Location: tmp/forki/facebook_media_ab664d32-4e47-4f5c-bf9e-c3a90330f97f.mp4
I, [2022-08-26T11:16:33.895209 #44745]  INFO -- : Size: 1.3282 MB
I, [2022-08-26T11:16:33.895246 #44745]  INFO -- : Time to Download: 0.345 seconds
I, [2022-08-26T11:16:58.776844 #44745]  INFO -- : Forki download started: https://scontent-dfw5-1.xx.fbcdn.net/v/t39.30808-1/249340936_460684055493192_7689068624385396462_n.png?stp=dst-png_p320x320&_nc_cat=1&ccb=1-7&_nc_sid=1eb0c7&_nc_ohc=2iBstN2J0zIAX-MX0Ln&_nc_ht=scontent-dfw5-1.xx&oh=00_AT8uE4Rei4zOh-l6wj0ARVhEjMQyiB46IkBtDFDwaN6mwg&oe=630EDDA3
I, [2022-08-26T11:16:58.819922 #44745]  INFO -- : Forki download finished
I, [2022-08-26T11:16:58.819965 #44745]  INFO -- : Save Location: tmp/forki/facebook_media_c4cebf3b-bd58-4f30-8db0-e88e9dc36624.png
I, [2022-08-26T11:16:58.819998 #44745]  INFO -- : Size: 0.018 MB
I, [2022-08-26T11:16:58.820016 #44745]  INFO -- : Time to Download: 0.043 seconds
/Users/justin/.rbenv/versions/3.1.2/lib/ruby/gems/3.1.0/bundler/gems/zorki-a53f7cf00e92/lib/zorki/scrapers/scraper.rb:71:in `get_content_of_subpage_from_url': Zorki could not find content requested (Zorki::ContentUnavailableError)
	from /Users/justin/.rbenv/versions/3.1.2/lib/ruby/gems/3.1.0/bundler/gems/zorki-a53f7cf00e92/lib/zorki/scrapers/user_scraper.rb:21:in `parse'
	from /Users/justin/.rbenv/versions/3.1.2/lib/ruby/gems/3.1.0/bundler/gems/zorki-a53f7cf00e92/lib/zorki/user.rb:46:in `block in scrape'
	from /Users/justin/.rbenv/versions/3.1.2/lib/ruby/gems/3.1.0/bundler/gems/zorki-a53f7cf00e92/lib/zorki/user.rb:45:in `map'
	from /Users/justin/.rbenv/versions/3.1.2/lib/ruby/gems/3.1.0/bundler/gems/zorki-a53f7cf00e92/lib/zorki/user.rb:45:in `scrape'
	from /Users/justin/.rbenv/versions/3.1.2/lib/ruby/gems/3.1.0/bundler/gems/zorki-a53f7cf00e92/lib/zorki/user.rb:12:in `lookup'
	from /Users/justin/.rbenv/versions/3.1.2/lib/ruby/gems/3.1.0/bundler/gems/zorki-a53f7cf00e92/lib/zorki/scrapers/post_scraper.rb:54:in `parse'
	from /Users/justin/.rbenv/versions/3.1.2/lib/ruby/gems/3.1.0/bundler/gems/zorki-a53f7cf00e92/lib/zorki/post.rb:40:in `block in scrape'
	from /Users/justin/.rbenv/versions/3.1.2/lib/ruby/gems/3.1.0/bundler/gems/zorki-a53f7cf00e92/lib/zorki/post.rb:39:in `map'
	from /Users/justin/.rbenv/versions/3.1.2/lib/ruby/gems/3.1.0/bundler/gems/zorki-a53f7cf00e92/lib/zorki/post.rb:39:in `scrape'
	from /Users/justin/.rbenv/versions/3.1.2/lib/ruby/gems/3.1.0/bundler/gems/zorki-a53f7cf00e92/lib/zorki/post.rb:8:in `lookup'
	from /Users/justin/Projects/duke/hypatia/app/media_sources/instagram_media_source.rb:54:in `retrieve_instagram_post'
	from /Users/justin/Projects/duke/hypatia/app/media_sources/instagram_media_source.rb:27:in `extract'
	from /Users/justin/Projects/duke/hypatia/test/media_sources/instagram_media_source_test.rb:6:in `<class:InstagramMediaSourceTest>'
	from /Users/justin/Projects/duke/hypatia/test/media_sources/instagram_media_source_test.rb:3:in `<main>'
	from /Users/justin/.rbenv/versions/3.1.2/lib/ruby/gems/3.1.0/gems/bootsnap-1.13.0/lib/bootsnap/load_path_cache/core_ext/kernel_require.rb:32:in `require'
	from /Users/justin/.rbenv/versions/3.1.2/lib/ruby/gems/3.1.0/gems/bootsnap-1.13.0/lib/bootsnap/load_path_cache/core_ext/kernel_require.rb:32:in `require'
	from /Users/justin/.rbenv/versions/3.1.2/lib/ruby/gems/3.1.0/gems/zeitwerk-2.6.0/lib/zeitwerk/kernel.rb:35:in `require'
	from /Users/justin/.rbenv/versions/3.1.2/lib/ruby/gems/3.1.0/gems/railties-7.0.3.1/lib/rails/test_unit/runner.rb:47:in `block in load_tests'
	from /Users/justin/.rbenv/versions/3.1.2/lib/ruby/gems/3.1.0/gems/railties-7.0.3.1/lib/rails/test_unit/runner.rb:47:in `each'
	from /Users/justin/.rbenv/versions/3.1.2/lib/ruby/gems/3.1.0/gems/railties-7.0.3.1/lib/rails/test_unit/runner.rb:47:in `load_tests'
	from /Users/justin/.rbenv/versions/3.1.2/lib/ruby/gems/3.1.0/gems/railties-7.0.3.1/lib/rails/test_unit/runner.rb:40:in `run'
	from /Users/justin/.rbenv/versions/3.1.2/lib/ruby/gems/3.1.0/gems/railties-7.0.3.1/lib/rails/commands/test/test_command.rb:33:in `perform'
	from /Users/justin/.rbenv/versions/3.1.2/lib/ruby/gems/3.1.0/gems/thor-1.2.1/lib/thor/command.rb:27:in `run'
	from /Users/justin/.rbenv/versions/3.1.2/lib/ruby/gems/3.1.0/gems/thor-1.2.1/lib/thor/invocation.rb:127:in `invoke_command'
	from /Users/justin/.rbenv/versions/3.1.2/lib/ruby/gems/3.1.0/gems/thor-1.2.1/lib/thor.rb:392:in `dispatch'
	from /Users/justin/.rbenv/versions/3.1.2/lib/ruby/gems/3.1.0/gems/railties-7.0.3.1/lib/rails/command/base.rb:87:in `perform'
	from /Users/justin/.rbenv/versions/3.1.2/lib/ruby/gems/3.1.0/gems/railties-7.0.3.1/lib/rails/command.rb:48:in `invoke'
	from /Users/justin/.rbenv/versions/3.1.2/lib/ruby/gems/3.1.0/gems/railties-7.0.3.1/lib/rails/commands.rb:18:in `<main>'
	from /Users/justin/.rbenv/versions/3.1.2/lib/ruby/gems/3.1.0/gems/bootsnap-1.13.0/lib/bootsnap/load_path_cache/core_ext/kernel_require.rb:32:in `require'
	from /Users/justin/.rbenv/versions/3.1.2/lib/ruby/gems/3.1.0/gems/bootsnap-1.13.0/lib/bootsnap/load_path_cache/core_ext/kernel_require.rb:32:in `require'
	from bin/rails:4:in `<main>'
```

</details>

**Pull request merged**